server/db/driver/pg/internal/matches.go

```diff
@@ -50,18 +50,30 @@ const (
 		aRedeemCoinID BYTEA,      -- the input spending the taker's contract output includes the secret
 		aRedeemTime INT8,         -- server time stamp
 		bSigAckOfARedeem BYTEA,   -- counterparty's (participant) sig with ack of initiator REDEEM data
+		bSigAckOfARedeemTime INT8,-- time at which the order is considered completed for the taker
 
 		-- participant/B (taker) REDEEM data
 		bRedeemCoinID BYTEA,
 		bRedeemTime INT8,         -- server time stamp
-		aSigAckOfBRedeem BYTEA    -- counterparty's (initiator) sig with ack of participant REDEEM data
+		aSigAckOfBRedeem BYTEA,   -- counterparty's (initiator) sig with ack of participant REDEEM data
+		aSigAckOfBRedeemTime INT8 -- time at which the order is considered completed for the maker
 	)`
 
+	RetrieveCompletedOrdersForAccount = `SELECT
+		takerOrder, takerAccount,
+		makerOrder, makerAccount,
+		bSigAckOfARedeemTime, aSigAckOfBRedeemTime -- a == maker, b == taker
+	FROM %s
+	WHERE (takerAccount = $1 AND bSigAckOfARedeemTime IS NOT NULL)  -- user is taker/b and they completed their part
+	   OR (makerAccount = $1 AND aSigAckOfBRedeemTime IS NOT NULL)  -- user is maker/a and they completed their part
+	ORDER BY bSigAckOfARedeemTime, aSigAckOfBRedeemTime
+	LIMIT $2; -- e.g. 25`
```

Comment:
Crap, this can't work as-is since (1) there can be multiple completed swaps for an order and thus orders will be repeated in the results, (2) the LIMIT does not make sense because of (1), and (3) nowhere is the actual order's status of the actual order considered and it could be in any state including booked or canceled.

One possible solution is to aggregate the unique orders in go code after scanning the results, but that doesn't solve the LIMIT or order status issues.

The query could be split into two, one for maker and one for taker, so that `DISTINCT ON (takerOrder)` and `DISTINCT ON (makerOrder)` may be used to only return 1 of a given order  id.  However, that does not solve (3) the fact that the orders could not be filled to completion.

All of this suggests that adding these redeem ack timestamps and using the matches table to identify completed orders was an ugly solution.  Maybe the orders table needs a completion time column added to make this straightforward.

Reply:
Resolved by adding a `complete_time` column to the orders tables and removing these matches changes.  Completed orders queries are simple now and involve the archive orders tables.

Further, it was necessary to only stamp and order with a completion time when the final swap involving the order completes, which involves checking not only order type/force/remaining but also the active swaps being handles by the Swapper.  Only the final swap should stamp the order as complete.

https://github.com/decred/dcrdex/pull/206/commits/994a33255262e624c35d64ced0ba3d564e0d6986 and https://github.com/decred/dcrdex/pull/206/commits/5042449a0831c8e2c4942c19f2f7c9cb4d914624 contain these changes.